### PR TITLE
Authentication model changes for remote access

### PR DESCRIPTION
**Pull request by n1v0lg**
This PR implements the necessary changes to the `Authentication` class, to support remote access authentication under the new remote cluster security model. Upon successful authentication, a new authentication instance will be constructed by the fulfilling cluster which combines information from the remote access API key used and the user authentication and role info sent by the querying cluster with a cross cluster request.

Remote access authentication is modeled in way that exposes (and assumes) that the underlying authentication method is an API key; for example, it includes the metadata associated with API keys in its metadata directly, re-using existing metadata field keys. I chose this approach instead of trying to generalize away from API keys because there are no medium-term plans to support any other authentication forms for remote access; generalizing would have made the change more complex.   

This change is stand-alone and not wired up to active code flows yet. A proof of concept in #92089 highlights how the model change in this PR fits into the broader context of the fulfilling cluster processing cross cluster requests.

### CODE REVIEW

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Authentication.java [hunk 1]
```diff
@@ -626,9 +647,18 @@ private void assertInternalConsistency() {
             || (getAuthenticatingSubject().getMetadata().get(AuthenticationField.API_KEY_ID_KEY) != null)
             : "API KEY authentication requires metadata to contain API KEY id, and the value must be non-null.";
 
+        assert (false == isAuthenticatedAsRemoteAccess())
+            || (getAuthenticatingSubject().getMetadata().get(AuthenticationField.API_KEY_ID_KEY) != null)
```
Comment (n1v0lg) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Authentication.java:651]: I opted for re-using the API key metadata fields as is (i.e., `REMOTE_ACCESS` authentication has an `API_KEY_ID_KEY` instead of a new `REMOTE_ACCESS_API_KEY_ID_KEY` field. This makes for a smaller change, and will make it easier to build roles in next PR (we can re-use the API key role descriptors parsing logic, for the API key specific role descriptor fields, for instance).

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Authentication.java [hunk 2]
```diff
@@ -882,6 +917,37 @@ public static Authentication newApiKeyAuthentication(AuthenticationResult<User>
...
+
+        final User remoteAccessApiKeyUser = remoteAccessApiKeyAuthcResult.getValue();
+        assert remoteAccessApiKeyUser.roles().length == 0
+            : "The user associated with a remote access API key authentication must have no role";
+
+        final User userFromRemoteCluster = authenticationFromRemoteCluster.getEffectiveSubject().getUser();
+        assert userFromRemoteCluster.enabled() : "The user received from a remote cluster must be enabled";
+
+        final User userWithoutRoles = new User(
+            userFromRemoteCluster.principal(),
```
Comment (n1v0lg) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Authentication.java:935]: I decided against combining the API key ID with the remote user principal: simple concatenation would require somewhat weird edge-case handling around max length restrictions (the API key ID together with an arbitrary username could exceed the limit); hashing would have been expensive and made the principal opaque. 

In terms of what appears in the logs and API responses, the remote user principal seems more useful than the API key ID which is why I've chosen it here. Furthermore, the API key ID is included in the realm name.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Subject.java [hunk 3]
```diff
@@ -99,6 +105,9 @@ public RoleReferenceIntersection getRoleReferenceIntersection(@Nullable Anonymou
                 return buildRoleReferencesForApiKey();
             case SERVICE_ACCOUNT:
                 return new RoleReferenceIntersection(new RoleReference.ServiceAccountRoleReference(user.principal()));
+            case REMOTE_ACCESS:
```
Comment (n1v0lg) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Subject.java:108]: Stay tuned for the next PR 🍿

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/xcontent/XContentUtils.java [hunk 4]
```diff
@@ -115,6 +115,12 @@ public static void addAuthorizationInfo(final XContentBuilder builder, final Map
                 builder.endObject();
             }
             case SERVICE_ACCOUNT -> builder.field("service_account", authenticationSubject.getUser().principal());
+            case REMOTE_ACCESS -> {
+                // TODO handle remote access authentication
```
Comment (n1v0lg) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/xcontent/XContentUtils.java:119]: Leaving this for when and if we deal with transforms, datafeeds etc

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/audit/logfile/LoggingAuditTrail.java [hunk 5]
```diff
@@ -1608,7 +1608,8 @@ private void setThreadContextField(ThreadContext threadContext, String threadCon
         LogEntryBuilder withAuthentication(Authentication authentication) {
             logEntry.with(PRINCIPAL_FIELD_NAME, authentication.getEffectiveSubject().getUser().principal());
             logEntry.with(AUTHENTICATION_TYPE_FIELD_NAME, authentication.getAuthenticationType().toString());
-            if (authentication.isApiKey()) {
+            // TODO more specialized auditing for remote access
```
Comment (n1v0lg) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/audit/logfile/LoggingAuditTrail.java:1611]: I'm planning to add a test for this as well -- just didn't get to it yet.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Authentication.java [hunk 6]
```diff
@@ -447,11 +466,25 @@ public void writeToContext(ThreadContext ctx) throws IOException, IllegalArgumen
     }
 
     public String encode() throws IOException {
+        return Base64.getEncoder().encodeToString(BytesReference.toBytes(toVersionedBytes()));
+    }
+
+    public BytesReference toVersionedBytes() throws IOException {
```
Comment (n1v0lg) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Authentication.java:472]: I'm blanking on a better name here (same for `fromVersionedBytes` below) but didn't want that to block the PR.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Authentication.java [hunk 7]
```diff
@@ -882,6 +929,37 @@ public static Authentication newApiKeyAuthentication(AuthenticationResult<User>
         return authentication;
     }
 
+    public static Authentication newRemoteAccessAuthentication(
+        AuthenticationResult<User> remoteAccessApiKeyAuthcResult,
+        Authentication authenticationFromRemoteCluster,
+        String nodeName
+    ) {
```
Comment (ywangd) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Authentication.java:936]: I'd personally prefer a different method signature that is not `static` and similar to the `token()` method, e.g.:
```java
public static toRemoteAccess(RemoteAccessAuthentication remoteAccessAuthentication ) {
  assert isApiKey(); 
  ...
}
```

I think it works better with composition of authenticatorChain. It will move some logic around remote authentication handling elsewhere into this method. But they feel like good changes to me.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Authentication.java [hunk 8]
```diff
@@ -366,6 +371,10 @@ public boolean isAuthenticatedAsApiKey() {
         return authenticatingSubject.getType() == Subject.Type.API_KEY;
     }
 
+    public boolean isAuthenticatedAsRemoteAccess() {
+        return authenticatingSubject.getType() == Subject.Type.REMOTE_ACCESS;
+    }
```
Comment (ywangd) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Authentication.java:376]: Nit: This method feels a bit odd to me. I understand it is modelled after existing methods but it just feels a bit off. I think the remote access authentication object will never have run-as. The authentication from QC can have run-as. We can even decide to support remote access API key run-as in future. But once these two authentication objects are combined, there should never be any possibility of run-as for the final authentication object. This means the method is redundant since we already have `isRemoteAccess` which reads better. I think a more suitable name is probably `isAuthenticatedWithRemoteAccessApiKey`. But now the semantic is different and requires a different implementation. Overall, I'd just drop this method if not absolutely necessary.

**Review by ywangd — COMMENTED** [review]
Haven't read the tests. But here are some comments for the main code changes.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Authentication.java (hunk 6 above)
Comment (n1v0lg) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Authentication.java:472]: I wanted to avoid the unnecessary base64 encoding for efficiency reasons (both the operation itself, and the slightly larger metadata blob). This is not critical though. In the interest of keeping things simple, I'll just re-use `encode` 👍

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Authentication.java [hunk 9]
```diff
@@ -882,6 +929,37 @@ public static Authentication newApiKeyAuthentication(AuthenticationResult<User>
...
+
+        final User userWithoutRoles = new User(
+            userFromRemoteCluster.principal(),
+            EMPTY_ARRAY,
+            userFromRemoteCluster.fullName(),
+            userFromRemoteCluster.email(),
+            userFromRemoteCluster.metadata(),
+            userFromRemoteCluster.enabled()
+        );
+        final Authentication.RealmRef authenticatedBy = newRemoteAccessRealmRef(remoteAccessApiKeyUser.principal(), nodeName);
```
[the remark this answers is not in the document]
Comment (n1v0lg) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Authentication.java:954]: Yes, absolutely. I'm a little confused as to why one of my tests passed that asserted the realm name consists of the API key ID but either way I'll look into the test and use the API key ID here. Thanks for catching this!

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Authentication.java (hunk 7 above)
Comment (n1v0lg) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Authentication.java:936]: I like your suggestion. Clean!

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Authentication.java (hunk 8 above)
Comment (n1v0lg) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Authentication.java:376]: Agreed, I'm all good to delete with method.

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/authc/AuthenticationTestHelper.java [hunk 10]
```diff
@@ -233,6 +237,43 @@ public static String randomInternalRoleName() {
         );
     }
 
+    public static RemoteAccessAuthentication randomRemoteAccessAuthentication() {
```
Comment (n1v0lg) [inline x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/authc/AuthenticationTestHelper.java:240]: The class name `RemoteAccessAuthentication` is a bit confusing, in retrospect. It's ambiguous if it refers to the header sent by the querying cluster, or a fulfilling-side `remote_access` authentication instance. I work on some renaming in a follow up PR (`RemoteAccessAuthenticationHeaders` to designate the headers sent by the QC or something along these lines).

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/ManageOwnApiKeyClusterPrivilege.java [hunk 11]
```diff
@@ -61,6 +61,7 @@ private ManageOwnClusterPermissionCheck() {
 
         @Override
         protected boolean extendedCheck(String action, TransportRequest request, Authentication authentication) {
+            authentication.assertNotRemoteAccess();
```
Comment (n1v0lg) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/ManageOwnApiKeyClusterPrivilege.java:64]: This is a cluster privilege, and the actions are all non-cross-cluster; as such it should never apply for remote access -- adding this assertion here as a sanity check. The real way to ensure nothing goes wrong here is most likely 1. enforcing a whitelist of request types that are allowed on the new remote access port, 2. excluding any non-index privileges from the role descriptors received in remote access headers.

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/authc/AuthenticationTestHelper.java [hunk 12]
```diff
@@ -421,10 +476,18 @@ public Authentication build(boolean maybeRunAsIfNotAlready) {
...
-                        authentication = Authentication.newApiKeyAuthentication(
+                        final Authentication apiKeyAuthentication = Authentication.newApiKeyAuthentication(
                             AuthenticationResult.success(user, metadata),
                             ESTestCase.randomAlphaOfLengthBetween(3, 8)
                         );
+                        // Remote access is authenticated via API key, but the underlying authentication instance has a different structure,
+                        // and a different subject type. If remoteAccessAuthentication is set, we transform the API key authentication
+                        // instance into a remote access authentication instance.
+                        // This means the builder only produces remote access authentication instances if `remoteAccess()` was called.
+                        // TODO generalize the builder to randomly build remote access authentication instance, when viable
```
Comment (n1v0lg) [inline x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/authc/AuthenticationTestHelper.java:487]: I thought I'd be able to generalize the builder in this PR, but this is more involved than I thought -- in a lot of places, it's actually not expected to have a remote_access authentication subject (e.g., anything unrelated to CCS for now); having `AuthenticationTestHelper.builder().build()` produce remote_access authentication instances would require updating quite a lot of tests and make the current PR less focused (also delaying it further). As such, I want to revisit this in a separate PR. Let me know if you think it's better to do this in one go instead.

**Comment by n1v0lg** [comment]
Docs failure is unrelated and already being investigated by the docs team

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/authc/AuthenticationTestHelper.java (hunk 10 above)
Comment (ywangd) [reply x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/authc/AuthenticationTestHelper.java:240]: Yeah I think rename might help. How about making it more specific to something like `QueryClusterAuthenticationInformation`?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/ManageOwnApiKeyClusterPrivilege.java (hunk 11 above)
Comment (ywangd) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/ManageOwnApiKeyClusterPrivilege.java:64]: I don't think we need this assertion here. Otherwise by analogy, we will need it for almost all cluster privileges. Why should this be special? 

> 1. enforcing a whitelist of request types that are allowed on the new remote access port,

We will need this. The sender side already has it. So similar check is needed on the receiver side since we know that it is possible to manually issue requests to the remote access port.

> 2. excluding any non-index privileges from the role descriptors received in remote access headers.

Isn't this already done in `RBACEngine.getRemoteAccessRoleDescriptorsIntersection`?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Authentication.java [hunk 13]
```diff
@@ -792,6 +828,11 @@ static RealmRef newApiKeyRealmRef(String nodeName) {
             // no domain for API Key tokens
             return new RealmRef(API_KEY_REALM_NAME, API_KEY_REALM_TYPE, nodeName, null);
         }
+
+        static RealmRef newRemoteAccessRealmRef(String apiKeyId, String nodeName) {
+            // no domain for remote access authentication
+            return new RealmRef(REMOTE_ACCESS_REALM_NAME_PREFIX + "_" + apiKeyId, REMOTE_ACCESS_REALM_TYPE, nodeName, null);
```
Comment (ywangd) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Authentication.java:834]: I made the suggestion of attaching API key ID to the realm name. But I am now questioning whether it is the right thing to do. It argubly makes the realm name look ugly. It leads to realm name explosion because each API key is now a diffferent realm. This could potentially cause issues for downstream processing, e.g. users want to aggregate information by realm name (the realm type is an alternative in some cases, but since you can have multiple realms of the same type, the alternative is not always viable).

I initially thought this could help ensure uniquness of users based on the underlying API keys. But now I think we don't really need it to achieve the same goal. Since we will have the API key ID in the Subject's metadata, uniqueness check should always be covered by, e.g:
* `canAccessResourceOf` - we will explicitly check the API key ID
* `equals`- this compares all metadata
* `hashCode` - this include all metadata as well

I think an analogy to this is the subject of an API key authentication. Without looking into the metadata for API key ID, two authentication of different API keys owned by the same user are essentially the same. But we do make sure checking into the metadata for uniqueness. So we don't have to rely on variation on the realm name. 

In summary, I now think we should just use a simple `_es_remote_access` for the realm name. What do you think? Sorry for the back and forth.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/RemoteAccessAuthentication.java [hunk 14]
```diff
@@ -124,6 +127,24 @@ public static RemoteAccessAuthentication decode(final String header) throws IOEx
...
+    public Map<String, Object> copyWithRemoteAccessEntries(final Map<String, Object> authenticationMetadata) {
+        assert false == authenticationMetadata.containsKey(AuthenticationField.REMOTE_ACCESS_AUTHENTICATION_KEY)
+            : "metadata already contains [" + AuthenticationField.REMOTE_ACCESS_AUTHENTICATION_KEY + "] entry";
+        assert false == authenticationMetadata.containsKey(AuthenticationField.REMOTE_ACCESS_ROLE_DESCRIPTORS_KEY)
+            : "metadata already contains [" + AuthenticationField.REMOTE_ACCESS_ROLE_DESCRIPTORS_KEY + "] entry";
+        final Map<String, Object> copy = new HashMap<>(authenticationMetadata);
+        try {
+            copy.put(AuthenticationField.REMOTE_ACCESS_AUTHENTICATION_KEY, getAuthentication().encode());
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
```
Comment (ywangd) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/RemoteAccessAuthentication.java:142]: I wonder whether we should just put it into the metadata without serialization. Serialization will still be done properly once the authentication object needs to go across wire. So storing the `authentication` object is not technically a problem. 

If we serialise it here, it needs to be deserialized again when needed, e.g. auditing, `canAccessResources` check. The callers also need to be aware of the desialization. That said both auditing and `canAccessResources` do not happen all the time. So the cost is not a lot. Having it serialised here makes deserialisation cheaper on another node (because it only needs to decode a string instead of `Authentication`). I don't see a clear winner here.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java [hunk 15]
```diff
@@ -195,6 +195,7 @@ public void authorizeClusterAction(
 
     // pkg private for testing
     static boolean checkSameUserPermissions(String action, TransportRequest request, Authentication authentication) {
+        authentication.assertNotRemoteAccess();
```
Comment (ywangd) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java:198]: Similarly here, I think it's better to have a allowlist check when receving the request instead of separate checks in downstream code.

**Review by ywangd — COMMENTED** [review]
A few more comments

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/RemoteAccessAuthentication.java (hunk 14 above)
Comment (ywangd) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/RemoteAccessAuthentication.java:142]: I just thought about a more valid argument for *not* serialising it here. The next node (of the FC) that recieves the overall authentication object could be at an older version than the current node. Therefore it may not be able to deserialise it if this is serialized using the current node's version. 

This is not possible today because we will prevent the overall authentication to be sent across at all if the other node is at an old version that does not understand `remote_access`. But it could be possible in future. Let's say 8.8 is the version that remote_access is introduced and some more changes to authentication is introduced in 8.9. Both QC and the initial node of FC is at 8.9, but FC is a mixed cluster that has 8.8 nodes, the authentication serialized here will not be parsable for the 8.8 nodes.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Authentication.java (hunk 13 above)
Comment (ywangd) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Authentication.java:834]: My initial thought of making realm name unique by appending API key ID is because we have this notion that "a user is uniquely identifiable by its principal and realm". This is true for realm users within a single cluster and we rely on that for things like user profiles. Therefore I was trying to apply the same principle for the remote access users. But this notion does not really work outside of realm users (and their OAuth2 tokens). For example, it does not work for API keys. It certainly does not work for users from different clusters. Even if we suffix the realm name with API key ID, it is still possible that the same API key is used by two different QCs and the two QCs happen to have the exact same user/realm. They are hence indistinguishable to FC. 

When we are going to implement `canAccessResourcesOf` for the new remote access authentication, I think we will have to establish that the user identity uniqueness across different clusters relies on each cluster having a different API key. FC Admins need to create and distribute the API keys properly to ensure resources are not accidentally shared between users from different clusters.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/ManageOwnApiKeyClusterPrivilege.java (hunk 11 above)
Comment (n1v0lg) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/ManageOwnApiKeyClusterPrivilege.java:64]: > Isn't this already done in RBACEngine.getRemoteAccessRoleDescriptorsIntersection?

It is, but that's on the querying cluster side -- we likely want this on the fulfilling cluster, to block manually crafted requests with unexpected privileges (e.g., cluster) from causing trouble.

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/authc/AuthenticationTestHelper.java (hunk 12 above)
Comment (n1v0lg) [reply x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/authc/AuthenticationTestHelper.java]: Nice, I'll take a look at `runAs` again and see if that works more cleanly. Otherwise will stick to the current approach. It's a bit clunky but works; since it's test code it'll be esp. easy to iterate on it.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/RemoteAccessAuthentication.java (hunk 14 above)
Comment (n1v0lg) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/RemoteAccessAuthentication.java:142]: Generally, if we can avoid serializing I'm all for it. However, I'm not positive this actually works -- this was a while ago now, but when I tried storing `Authentication` directly in metadata as part of the fulfilling cluster POC, requests failed with serialization errors. I'll see if I missed something last time. I'm ++ on your suggestion, if it works as expected.

Comment (ywangd) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/RemoteAccessAuthentication.java:142]: I think it might be because Authentication does not implement Writable. I don't see a reason why it cannot since its writeTo method is compatible with the interface. It would be great if you could take a closer look. Thanks!

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Authentication.java (hunk 13 above)
Comment (n1v0lg) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Authentication.java:834]: Thanks for thinking through this thoroughly, and apologies for not spotting this myself. I agree that `_es_remote_access_{ID}` may complicate things for end-users, e.g., if they want to agg by the `_es_remote_access` realm name. At the same time, they might actually want to agg by different API keys, that is aggregating by `_es_remote_access_{ID}` is exactly what they want. It's a little hard for me to tell, but I'm leaning towards your assessment that users will more often prefer to aggregate (or search) by `_es_remote_access`. That naming is also more in line with other synthetic realms, so I'm in favor of just `_es_remote_access` as the realm name.

> When we are going to implement canAccessResourcesOf for the new remote access authentication, I think we will have to establish that the user identity uniqueness across different clusters relies on each cluster having a different API key.

This is a really good point. My mental model here is that we shouldn't necessarily think of clusters as the authenticated entity, but the API key itself (or whatever other credential we might use in the future). Anything with access to the API key is treated as a single entity, from the point of view of the FC. So if we have two different querying clusters with users with overlapping principals they should be treated as one entity if the clusters use the same API key, and different entities if they use different keys. This is a potential gotcha for administrators and we'll have to be clear about this in the docs, but to me this is the only meaningful model for resource sharing. It's somewhat analogous to two different users sharing credentials to a single account -- there is nothing to be done about this, other than establishing that sharing passwords and other credentials is bad practice.

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/authc/AuthenticationTestHelper.java (hunk 10 above)
Comment (n1v0lg) [reply x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/authc/AuthenticationTestHelper.java:240]: I'm thinking `RemoteClusterAuthenticationInformation` to avoid Query or QueryingCluster -- to me these terms are largely for internal use and even though they've made it into PR descriptions and some test code I slightly prefer sticking to RemoteCluster in prod code, unless it's really ambiguous.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/RemoteAccessAuthentication.java (hunk 14 above)
Comment (n1v0lg) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/RemoteAccessAuthentication.java:142]: Ah wait, I think I missed part of your point: if we store `authentication` as bytes in metadata, we'd have to deal with rewriting for an older node at two stages: rewrite the outer authentication instance, and also explicitly deserialize, rewrite, and serialize again the instance included in metadata. 

I'll have to think about this some more.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Authentication.java [hunk 16]
```diff
@@ -934,6 +1052,39 @@ private static Map<String, Object> maybeRewriteMetadataForApiKeyRoleDescriptors(
         return metadata;
     }
 
+    // pkg-private for testing
+    static Map<String, Object> maybeRewriteMetadataForRemoteAccessAuthentication(
```
Comment (n1v0lg) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Authentication.java:1056]: This is based on part of our discussion around BWC in your review comment. 

The rewrite logic here is provisional: for 8.7, we won't ever end up rewriting since remote access authentication cannot be rewritten towards anything older than 8.7. However, it seems prudent to have the logic in place now -- it's a loaded foot gun otherwise as we move into versions past 8.7, and possibly change the authentication model. I'm okay also holding off with introducing this in 8.8, as it makes testing less awkward. Let me know what you prefer.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/RemoteAccessAuthentication.java (hunk 14 above)
Comment (n1v0lg) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/RemoteAccessAuthentication.java:144]: Let's sync on whether to encode or not once you're back -- I'm going with encoding for now because working around `writeGenericMap` is not so straight-forward. 

Regarding your point around needing to rewrite the authentication instance that's stored inside metadata, I've added code addressing it.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Authentication.java [hunk 17]
```diff
@@ -882,11 +973,38 @@ public static Authentication newApiKeyAuthentication(AuthenticationResult<User>
...
+        assert false == authentication.isAssignedToDomain();
+        return authentication;
+    }
+
     private static RealmRef maybeRewriteRealmRef(Version streamVersion, RealmRef realmRef) {
         if (realmRef != null && realmRef.getDomain() != null && streamVersion.before(VERSION_REALM_DOMAINS)) {
             logger.info("Rewriting realm [" + realmRef + "] without domain");
             // security domain erasure
-            new RealmRef(realmRef.getName(), realmRef.getType(), realmRef.getNodeName(), null);
+            return new RealmRef(realmRef.getName(), realmRef.getType(), realmRef.getNodeName(), null);
```
Comment (n1v0lg) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Authentication.java:1007]: Fixing this in a separate PR: https://github.com/elastic/elasticsearch/pull/93276 -- will merge 93276 first.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/ManageOwnApiKeyClusterPrivilege.java (hunk 11 above)
Comment (ywangd) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/ManageOwnApiKeyClusterPrivilege.java]: Right, we always need pair the checks on both sides because the requests can be manually crafted.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/RemoteAccessAuthentication.java (hunk 14 above)
Comment (ywangd) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/RemoteAccessAuthentication.java:144]: Thanks for looking into it. TIL. We might want to revisit how this can be done without encoding before initial release. But I agree for now let's just go with encoding.

**Review by ywangd — APPROVED** [review]
LGTM

As discussed we can revisit the writeGenericMap support for `Authentication` in future (preferably before initial release so we don't need to handle yet another BWC). 

I think we also want to reconcile where we handle version check/rewrite for the Authentication object. Right now it is handled in both `writeTo` and `maybeRewriteForOlderVersion`. When Albert and I were at it previously, we sorta agreed that the version should only be handled at serialization time, i.e. `writeTo`. This is still an open item. Obviously it does not really belong to this PR. But the additional version handling added does prompt me to think about it. So figured I'd just mention it for the record. Also the method name `maybeRewriteForOlderVersion` is technically wrong because it can actually write into a newer version, i.e. the `olderVersion` maybe newer than the authentication's version ...

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Authentication.java [hunk 18]
```diff
@@ -938,6 +1055,39 @@ private static Map<String, Object> maybeRewriteMetadataForApiKeyRoleDescriptors(
         return metadata;
     }
 
+    // pkg-private for testing
+    static Map<String, Object> maybeRewriteMetadataForRemoteAccessAuthentication(
+        final TransportVersion olderVersion,
+        final Authentication authentication
+    ) {
+        // TODO also handle remote access authentication executed with old API keys (either by rejecting this, or rewriting role
+        // descriptors)
```
[the remark this answers is not in the document]
Comment (n1v0lg) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Authentication.java:1064]: It's a weird edge case but technically nothing stands in the way of an admin taking an old API key that has something like `superuser` permissions and handing that to a querying cluster to use as a remote access key. I think it's fair for us to prevent this with a version check on the API key during authentication but it's worth failing gracefully

[the remark this answers is not in the document]
Comment (ywangd) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Authentication.java:1064]: I may be missing something. Why is an old API key from FC with `superuser` permission a problem when it is used as a remote access key? It has more privileges than necessary but that is not an unique problem for old API keys. The creation version of the API key is *not* the same as the authentication's version. The authentication's version should always be `CURRENT` from a successful authentication process. As such, the role descriptors should be stored as bytes even if the API key itself is old.

[the remark this answers is not in the document]
Comment (n1v0lg) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Authentication.java:1064]: You're right, I had my wires crossed: to paraphrase, since we only introduced remote access in 8.7, by the time we get to the rewrite method, (remote access) authentication based on an old (map-based role descriptors) API key will always already have the role descriptors written in its metadata as bytes, and we don't need to worry about having to rewrite them here. 

I will remove the nonsensical TODO 👍

**Comment by n1v0lg** [comment]
@elasticmachine update branch

**Comment by n1v0lg** [comment]
@elasticmachine run elasticsearch-ci/part-1-fips
